src/emmy/expression/compile.cljc

```diff
@@ -481,53 +669,23 @@
        (str "`compile-fn` can only infer arity for functions with just one
            arity, not " arity ". Please pass an explicit `n`.")))))
 
-(defn compile-fn*
+(defn compile-fn
   "Returns a compiled, simplified version of `f`, given a function `f` of arity
   `n` (i.e., able to accept `n` symbolic arguments).
 
   `n` defaults to `([[f/arity]] f)`.
 
+  You may also specify options in the third argument. See [[compile-state-fn]]
+  for information on the options supported.
+
   The returned, compiled function expects `n` `Double` (or `js/Number`)
   arguments. The function body is simplified and all common subexpressions
-  identified during compilation are extracted and computed only once.
-
-  NOTE: this function uses no cache. To take advantage of the global compilation
-  cache, see `compile-fn`."
-  ([f] (compile-fn* f (retrieve-arity f)))
+  identified during compilation are extracted and computed only once."
+  ([f] (compile-fn f (retrieve-arity f)))
   ([f n]
-   (let [sw       (us/stopwatch)
-         args     (repeatedly n #(gensym 'x))
-         body     (-> (apply f args)
-                      (g/simplify)
-                      (v/freeze)
-                      (cse))
-         compiled (case (compiler-mode)
-                    :source (#?(:clj compile->clj :cljs compile->js) args body)
-                    :js (compile->js args body)
-                    :clj (compile->clj args body)
-                    :native (compile-native args body)
-                    :sci (compile-sci args body))]
-     (log/info "compiled function of arity" n "in" (us/repr sw) "with mode" *mode*)
-     compiled)))
-
-(defn compile-fn
-  "Memoized version of [[compile-fn*]]. See that function's docs for more detail.
-
-  NOTE: that this function makes use of a global compilation cache, keyed by the
-  vector `[f n *mode*]`. See `compile-fn*` to avoid the cache."
-  ([f] (let [[kwd n :as arity] (f/arity f)]
-         (when-not (= kwd :exactly)
-           (u/illegal
-            (str "`compile-fn` can only infer arity for functions with just one
-           arity, not " arity ". Please pass an explicit `n`.")))
-         (compile-fn f n)))
-  ([f n]
-   (let [mode *mode*]
-     (if-let [cached (@fn-cache [f n mode])]
-       (do
-         (log/info "compiled function cache hit - arity " n ", mode " mode)
-         cached)
-       (binding [*mode* mode]
-         (let [compiled (compile-fn* f n)]
-           (swap! fn-cache assoc [f n mode] compiled)
-           compiled))))))
+   (compile-fn f n {}))
+  ([f n opts]
+   (let [argv (into [] (repeatedly n #(gensym 'x)))]
```

Comment:
If we make parameters a keyword option, do you think we could merge `compile-fn` and `compile-state-fn` into a single thing? `compile-state-fn` would be a `compile-fn` call with parameters supplied, otherwise the function's treated as one level deep?